### PR TITLE
chore(snippets): generate snippets for helpers

### PR DESCRIPTION
**Pull request by millotp**
## 🧭 What and Why

Generate snippets for helpers, we might want to add `isSnippet` too to only select the best looking one, but for now its fine.
I had to refacto the `SnippetsGenerator` and stop sharing code with the `TestsRequest` class because it was too messy with too many branches.

### CODE REVIEW

File: generators/src/main/java/com/algolia/codegen/cts/tests/SnippetsGenerator.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+      new SupportingFile(
+        "snippets/method.mustache",
+        "snippets/" + language + outputFolder + Helpers.createClientName(client, language) + extension
+      )
+    );
+  }
+
+  private Map<String, Snippet[]> loadSnippets() throws Exception {
+    String clientName = client;
+    // This special case allow us to read the `search` CTS to generated the tests for the
```
Comment (shortcuts) [inline generators/src/main/java/com/algolia/codegen/cts/tests/SnippetsGenerator.java:52]:
```suggestion
    // This special case allow us to read the `search` CTS to generated the snippets for the
```

File: generators/src/main/java/com/algolia/codegen/cts/tests/SnippetsGenerator.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    if (client.equals("algoliasearch")) {
+      clientName = "search";
+    }
+
+    Map<String, Snippet[]> baseSnippets = loadCTS("requests", clientName, Snippet[].class);
+
+    // The algoliasearch client bundles many client and therefore should provide snippets for all
+    // the subsequent specs
+    if (client.equals("algoliasearch")) {
+      Map<String, Snippet[]> recommendSnippets = loadCTS("requests", "recommend", Snippet[].class);
```
Comment (shortcuts) [inline generators/src/main/java/com/algolia/codegen/cts/tests/SnippetsGenerator.java:63]: just realized but we don't do analytics etc. here?

**Review by shortcuts — COMMENTED** [review]
this looks wonderful, small question about algoliasearch

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsGenerator.java [hunk 3]
```diff
@@ -57,6 +58,38 @@ protected <T> Map<String, T> loadCTS(String path, String clientName, Class<T> js
     return cts;
   }
 
+  protected <T> Map<String, T[]> loadFullCTS(Class<T[]> jsonType) throws Exception {
+    String clientName = client;
+    // This special case allow us to read the `search` CTS to generated the tests for the
+    // `lite` client, which is only available in Javascript
```
Comment (shortcuts) [inline generators/src/main/java/com/algolia/codegen/cts/tests/TestsGenerator.java:64]: It's in dart too no?

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsGenerator.java [hunk 4]
```diff
@@ -57,6 +58,38 @@ protected <T> Map<String, T> loadCTS(String path, String clientName, Class<T> js
...
+    if (client.equals("algoliasearch")) {
+      clientName = "search";
+    }
+
+    Map<String, T[]> baseCTS = loadCTS("requests", clientName, jsonType);
+
+    // The algoliasearch client bundles many client and therefore should provide tests for all the
```
Comment (shortcuts) [inline generators/src/main/java/com/algolia/codegen/cts/tests/TestsGenerator.java:71]:
```suggestion
    // The lite client bundles many client and therefore should provide tests for all the
```

**Review by shortcuts — COMMENTED** [review]
Looks awesome! I only have nitpick comments so non blocking, will check the gen

**Comment by shortcuts** [comment]
Gen looks incredible

File: generators/src/main/java/com/algolia/codegen/cts/tests/SnippetsGenerator.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        "snippets/" + language + outputFolder + Helpers.createClientName(client, language) + extension
+      )
+    );
+  }
+
+  private Map<String, Snippet[]> loadSnippets(Map<String, CodegenOperation> operations) throws Exception {
+    Map<String, Snippet[]> snippets = loadFullCTS(Snippet[].class);
+
+    String clientName = client;
+    if (client.equals("algoliasearch")) {
```
[the remark this answers is not in the document]
Comment (millotp) [reply generators/src/main/java/com/algolia/codegen/cts/tests/SnippetsGenerator.java:54]: yep agree

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsGenerator.java (hunk 3 above)
Comment (millotp) [reply generators/src/main/java/com/algolia/codegen/cts/tests/TestsGenerator.java:64]: true, this is an old comment

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsGenerator.java (hunk 4 above)
Comment (millotp) [reply generators/src/main/java/com/algolia/codegen/cts/tests/TestsGenerator.java:71]: I'll do all the renaming in another PR

**Comment by millotp** [comment]
> Can we make the generator replace the values of indexName params with something like YOUR_INDEX_NAME?

I tried a little hack to make this work, assuming the parameter is always called `indexName`, it's already way better !

**Review by shortcuts — APPROVED** [review]
💅🏻 CLEANNNN

**Pull request merged**